pkg/operator/ceph/object/objectstore.go

```diff
@@ -1010,3 +1017,31 @@ func errorOrIsNotFound(err error, msg string, args ...string) error {
 	}
 	return errors.Wrapf(err, msg, args)
 }
+
+// Update the Endpoints in Zone with new details
+func updateZoneEndpointList(zoneEndpointsList []string, oldEndpoint, newEndpoint string) []string {
+	appendNewEndpoint := false
+	if oldEndpoint != newEndpoint {
+		for i, endpoint := range zoneEndpointsList {
+			if oldEndpoint == endpoint {
+				zoneEndpointsList = append(zoneEndpointsList[:i], zoneEndpointsList[i+1:]...)
+				break
+			}
+		}
+		appendNewEndpoint = true
+	} else {
+		for i, endpoint := range zoneEndpointsList {
+			if newEndpoint == endpoint {
+				break
+			}
+			if i == len(zoneEndpointsList) {
+				appendNewEndpoint = true
+			}
+		}
+	}
+
+	if appendNewEndpoint {
+		zoneEndpointsList = append(zoneEndpointsList, newEndpoint)
+	}
+	return zoneEndpointsList
+}
```

Comment:
I think I was looking a bit to hard at **what** was happening instead of **why** it was happening, and now I have questions. In [this comment](https://github.com/rook/rook/pull/9622#issuecomment-1078434303), I mentioned that Ali and I had a meeting and determined that the endpoint should be set to a singular (not plural) value. Why are we not merely setting a singular endpoint in the RGW here? Am I missing something? @alimaredia do you recall this conversation and have thoughts?

Reply:
Chatting with Ali, I think I understand now: each CephObjectStore needs to add its own endpoint into this list. So it is necessary to do some merging.

Reply:
The problem with appending endpoints instead of setting them explicitly is if one converts a single-site to multi-site, the existing service IP will persist.  Also, there's no good reason to have multiple ObjectStores pointing to the same zone, and even if there was, if one ObjectStore goes away, that entry will also be abandoned.  All these scenarios will break replication.

Reply:
The concept of the zone includes a set of RGW servers using the same osd pool IMO the same relationship applies for Zone CR and CephObjectStore CR.  We can add multiple instances of RGW in cephobjectstore CR, but all the servers have same k8s service endpoints. If the user needs to have different endpoints for each RGW server then he need to define different cephobjectstore CR. Above if rgw server is faulty then if it breaks the replication, that applies for normal ceph multisite set up as well right?

Reply:
@BlaineEXE can we please get rid of the merging logic.  It's only going to break users, and it will help no one.  1 zone should have 1 endpoint, this endpoint is only used for replication purposes.

So, for non-multisite:  Don't set any endpoint.
For multi-site: set service IP by default, otherwise set value specified in spec.

There should be n=1 values for endpoint for each zone.

Reply:
I'm on the fence about this. 

I do see some value to considering Jiffin's response in regards to what happens if a user wants multiple ingress into an RGW cluster.

On one hand, wanting redundant ingress seems silly since any k8s user should be able to easily configure a load balancer to the RGW cluster, eliminating the need for redundant ingress.

On another hand, I could see some benefit to having multiple (but not necessarily redundant) ingress. Perhaps a user would want to have two different "classes" of RGW clusters. One with beefy RGWs that can handle a lot of client throughput, and another cluster with smaller RGWs that give clients fast-enough access. But I'm not sure how much that scenario might realistically be desired in the real world. For a user with needs that advanced, I suspect they might want standalone ceph servers.

@cbodley or @alimaredia do you care to weigh in here?

Reply:
i don't really see a problem with having multiple Stores per Zone

some multisite customers have found it useful to dedicate some radosgw nodes in each zone just to client traffic and others just to replication, with separate load balancers for each. would two Store CRs be the easiest way to manage something like that in rook? in this case, you'd only want one of them added to the Zone endpoints

but there may be other reasons to have multiple Stores, unrelated to multisite. if you *don't* want any particular Store dedicated to replication, it does still make sense to add each of them to the Zone endpoints

this merging is probably the right default behavior, but we might want an option in the Store CR that can hide its endpoint from the Zone

Reply:
> @mgugino-uipath: if one ObjectStore goes away, that entry will also be abandoned.

if the ObjectStore is removed, or its endpoint changes, it's important that these zone endpoints get updated accordingly. is that logic missing?

the endpoint-merging logic here would be fine, if those other events were handled correctly

Reply:
There's just no good reason to merge the endpoints.  If the Objectstore is removed, presumably another one will take it's place, or the zone is going away.  In either case, no merging needed.  Replication will not try multiple endpoints, it will pick one (the first one?) and if that one doesn't work, replication is broken.  There is no reason to have multiple endpoints (ceph itself accepts multiple endpoints, possibly should not).

Reply:
>  Replication will not try multiple endpoints, it will pick one (the first one?) and if that one doesn't work, replication is broken.

@mgugino-uipath rgw sends http requests to these endpoints in round-robin fashion, see https://github.com/ceph/ceph/blob/f38cdd9c/src/rgw/rgw_rest_conn.cc#L68-L79

if one of these endpoints is down, replication handles that like any other error and keeps retrying with different endpoints

> If the Objectstore is removed, presumably another one will take it's place, or the zone is going away.

it sounds like you want to require a 1-to-1 mapping between ObjectStores and Zones, but @thotz and i both described use cases for multiple ObjectStores. are there better ways to accomplish these in a single ObjectStore CR?

Reply:
@cbodley I'm not sure that round-robin code actually works.  I know if I set the endpoints by hand, then the operator reconciles again and adds the service IPs, replication does not work.  If you try to run a sync via radosgw-admin, you'll get timeouts indefinitely.  This happens for me on 16.x of ceph, other releases I'm not as sure.

> it sounds like you want to require a 1-to-1 mapping between ObjectStores and Zones

Strictly speaking, the replication endpoint belongs to the zone object IMO, but the current code base makes the ObjectStore much easier to implement (especially if using service IPs, the objectstore and it's associated svc won't be created yet, if for some reason you replicate zones within a single k8s cluster).

Anyway, what I care about is a 1-to-1 mapping between zones and replication endpoints, you can have as many objectstores as you wish so long as only 1 sets a replication endpoint.

Reply:
The breaking behavior of the service IP's is also confirmed by this user: https://github.com/rook/rook/issues/6432#issuecomment-923820650

So, for whatever reason, the round robin isn't working.

Reply:
round-robin (and associated error handling) works and is well-tested. it's just that some radosgw-admin commands (`sync status`, `realm pull`, `bucket sync run`) don't have any retry behavior built-in, so will fail if any endpoints can't be reached

so we shouldn't leave bad endpoints in this list. it's perfectly fine to have 2 or more good endpoints

Reply:
How do you explain replication ceasing to work for myself and that other user when the service IP's are re-added?  Data stops getting replicated, I have tested this by uploading data to one cluster, and the data never appears on the other.